### PR TITLE
docs for components sharing state

### PR DESCRIPTION
**Pull request by acivitillo**
Adding documentation for components sharing state as discussed here: https://github.com/idom-team/idom/discussions/561

I have added the 2 classic React examples: 1) synced input; 2) filterable list.

### CODE REVIEW

**Comment by rmorshea** [comment]
I'll take a look at this shortly. Thanks for the contribution!

File: docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  {
+    "name": "Shish Kebab",
+    "description": "Shish kebab is a popular meal of skewered and grilled cubes of meat",
+  },
+  {
+    "name": "Dim sum",
+    "description": "Dim sum is a large range of small dishes that Cantonese people traditionally enjoy in restaurants for breakfast and lunch",
+  }
+]
```
Comment (rmorshea) [inline docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json:22]: Can you add a new line here?

**Comment by rmorshea** [comment]
Ultimately I think this content belongs under the "Managing State" section, but since it will probably be some time before I get around to writing more documentation, I'm going to accept this and move these docs around later. Thanks for writing this up!

File: docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json (hunk 1 above)
Comment (rmorshea) [inline docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json:22]:
```suggestion
]

```

File: docs/source/adding-interactivity/components-sharing-state/index.rst [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    @component
+    def Input(label, value, set_value):
+        def handle_change(event):
+            set_value(event["target"]["value"])
+
+        return html.label(
+            label + " ", html.input({"value": value, "onChange": handle_change})
+        )
+
+    run(SyncedInputs)
```
Comment (rmorshea) [inline docs/source/adding-interactivity/components-sharing-state/index.rst:42]: Can this be made into an interactive example in the same way you did for `filterable_list`?

File: docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
+[
+  {
+    "name": "Sushi",
+    "description": "Sushi is a traditional Japanese dish of prepared vinegared rice",
```
Comment (rmorshea) [inline docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json:4]:
```suggestion
    "description": "Sushi is a traditional Japanese dish of prepared vinegared rice"
```

File: docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  {
+    "name": "Dal",
+    "description": "The most common way of preparing dal is in the form of a soup to which onions, tomatoes and various spices may be added",
```
Comment (rmorshea) [inline docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json:8]:
```suggestion
    "description": "The most common way of preparing dal is in the form of a soup to which onions, tomatoes and various spices may be added"
```

File: docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  {
+    "name": "Pierogi",
+    "description": "Pierogi are filled dumplings made by wrapping unleavened dough around a savoury or sweet filling and cooking in boiling water",
```
Comment (rmorshea) [inline docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json:12]:
```suggestion
    "description": "Pierogi are filled dumplings made by wrapping unleavened dough around a savoury or sweet filling and cooking in boiling water"
```

File: docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  {
+    "name": "Shish Kebab",
+    "description": "Shish kebab is a popular meal of skewered and grilled cubes of meat",
```
Comment (rmorshea) [inline docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json:16]:
```suggestion
    "description": "Shish kebab is a popular meal of skewered and grilled cubes of meat"
```

File: docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  {
+    "name": "Dim sum",
+    "description": "Dim sum is a large range of small dishes that Cantonese people traditionally enjoy in restaurants for breakfast and lunch",
```
Comment (rmorshea) [inline docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/data.json:20]:
```suggestion
    "description": "Dim sum is a large range of small dishes that Cantonese people traditionally enjoy in restaurants for breakfast and lunch"
```

File: docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/app.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+    return html.label(
+        "Search by Food Name: ", html.input({"value": value, "onChange": handle_change})
+    )
+
+
+@component
+def Table(value, set_value):
+    rows = []
+    for row in data:
```
Comment (rmorshea) [inline docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/app.py:31]:
```suggestion
    for row in food_data:
```

File: docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/app.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
+import json
+from pathlib import Path
+
+from idom import component, hooks, html, run
+
+
+HERE = Path(__file__)
+DATA_PATH = HERE.parent / "data.json"
+sculpture_data = json.loads(DATA_PATH.read_text())
```
Comment (rmorshea) [inline docs/source/adding-interactivity/components-sharing-state/_examples/filterable_list/app.py:9]:
```suggestion
food_data = json.loads(DATA_PATH.read_text())
```

**Review by rmorshea — CHANGES_REQUESTED** [review]

**Comment by acivitillo** [comment]
thanks for the quick review! I will work on it in the weekend

File: docs/source/adding-interactivity/components-sharing-state/index.rst (hunk 2 above)
Comment (acivitillo) [reply docs/source/adding-interactivity/components-sharing-state/index.rst:42]: of course

**Comment by acivitillo** [comment]
OK I created the additional example for syncedinputs and moved the docs to managing-state

**Comment by rmorshea** [comment]
It looks like there's just a minor style issue. If you're got your [dev environment](https://idom-docs.herokuapp.com/docs/developing-idom/contributor-guide.html#development-environment) set up you should be able to just run `isort .` to fix the issue.

*Also, let me know if you experienced any issues contributing to IDOM while working on Windows. I have a Windows machine I can use to test things out myself so it's very possible there are gaps in the contributor guide.*

**Comment by acivitillo** [comment]
ok I fixed the sorting of imports in the examples, it should work now

**Comment by acivitillo** [comment]
I run the `nox -s test_docs`. Sphinx passes but then I get a fail on docker it seems, I don't understand this part.

image

**Comment by acivitillo** [comment]
OK I had to fix 2 things to make this work and `nox -s test_docs` should now pass.

1) I had to upgrade npm in `Dockerfile` and `fast-json-path` in `docs/source/_cusom_js/package-lock.json` and in `src/client/package-lock.json`

2) I had to uncomment a 2nd Docker build in `noxfile.py`. nox is building the exact same docker image 2 times, from the looks of it the second build is not necessary and causes a bug (explained below)

Very interesting bug. The `test_docs` is building the same docker image 2 times, fix 1 above fixed the first build, but still the second build broke. The error is "Could not load /app/src/client/node_modules/fast-json-patch/index.mjs". This seems to happen because during the first docker build a folder `node_modules` is created inside the folder `/client` and for some reason this breaks the build. In fact, when I manually delete that folder it works.

File: noxfile.py [hunk 10]
```diff
@@ -248,7 +248,16 @@ def test_docs(session: Session) -> None:
...
+    # session.run(
+    #    "docker",
+    #    "build",
+    #    ".",
+    #    "--file",
+    #    "docs/Dockerfile",
+    #    "--tag",
+    #   "idom-docs:latest",
+    #    external=True,
+    # )
```
Comment (rmorshea) [inline noxfile.py:260]: I'm not quite sure what's going wrong here. Were you seeing problems with this locally or in CI?

Comment (acivitillo) [reply noxfile.py:260]: I got the same error in local and in CI: `Could not load /app/src/client/node_modules/fast-json-patch/index.mjs`. 

But take a look at line `124`, there is another `docker build` in the `noxfile.py`, so `nox` is building the same docker image 2 times. 

I think the first time `docker build` creates `node_modules` folder in `client` and for some reason, this breaks the second docker build.

Comment (rmorshea) [reply noxfile.py:260]: I'm gonna see what happens if I uncomment this. If it breaks in CI we can investigate.

Comment (rmorshea) [inline noxfile.py:260]:
```suggestion
    session.run("docker", "build", ".", "--file", "docs/Dockerfile", external=True)
```

Comment (rmorshea) [reply noxfile.py]: Looks to be working now.

**Review by rmorshea — APPROVED** [review]

File: noxfile.py (hunk 10 above)
Comment (acivitillo) [reply noxfile.py]: OK, then I will need to figure out how to run it locally w/o breaking. So the actual fix here was upgrading npm and fast-json-patch it seems.

**Pull request merged**